_posts/2020-12-22-aws-antipatterns.md

```diff
@@ -0,0 +1,198 @@
+- Creating virtual servers, it's possible to choose their type and capacity, and some of them can be very expensive if they are left running. Machine learning projects might often run these servers, but are typically not supposed to run them 24/7, only on demand.
+- Sagemaker developer endpoints can also rack up some bill quite fast, it's easy to create them and use them, but if you forget to shut them down, euros will start accumulating rapidly
+- There are often few ways to model things, and the cost may vary quite a lot. Different patterns depending on your case.
+- If you're not seeing how much things cost, you might be enticed to create a new feature for the product, based on some new service just because you can. Perhaps it's something you wouldn't have even considered before.
+
+This can often be a problem because I've seen that billing is often abstracted, denied, and hidden from the development team. So they can freely 
+
+So the fix is rather simple: Create visibility for the costs. Use billing and costs as one input for your architecture, from the very beginning. This allows applying one AWS Architects centric skill: Cost optimization. This is again one of those things that can be expensive to fix afterward, it's cheapest to build in cost optimization from the beginning. Having this visibility can help you make educated choices between different architectures. What you're looking for is a steady growth of expenses as your resource usage grows, but any dramatic anomalies or changes in the bills should immediately be studied and understood.
+
+Highly recommended also to enable billing alerts, so that you get a notification if you are going over budgeted amounts. Early warning, so you don't need to study the reports every day. When you get a warning, you check if the change is valid, aka in line with improved architecture and increased value. Or if the change was unexpected, a result of a bad pattern or badly scaling architecture. Even experienced cloud developers have sometimes been surprised by the bills.
+
+So: Keep that agility and velocity - but also keep your eyes open. Do also cost-driven development, in addition to all your other viewpoints. Don't do things just because you can, think about costs, benefits, and ROI. That way you can make your development less eventful, but much more peaceful. Fewer surprises equal more happiness, in this case.
+
+
+## Conclusion
+
+For each antipattern, there's a cure. If you build it in your development process, from the very beginning, life will be much better for you. Some of these are impossible to solve on very high and abstract levels, so are best taken care of by empowered DevOps teams that want to carry the responsibility of not only building the services but making sure they run smoothly as well.
```

Comment:
>there's a cure
 Maybe a mitigation? Somehow "cure" comes across as related to a health condition :-D

Reply:
Yeah, I like that more too.